rcl/include/rcl/parameter_client.h

```diff
@@ -0,0 +1,185 @@
+// Copyright 2016 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCL__PARAMETER_CLIENT_H_
+#define RCL__PARAMETER_CLIENT_H_
+
+#if __cplusplus
+extern "C"
+{
+#endif
+
+#include <rcl_interfaces/msg/list_parameters_result__struct.h>
+#include <rcl_interfaces/msg/parameter__struct.h>
+#include <rcl_interfaces/msg/parameter_event__struct.h>
+#include <rcl_interfaces/msg/set_parameters_result__struct.h>
+
+#include "rcl/client.h"
+#include "rcl/node.h"
+#include "rcl/wait.h"
+
+/// TODO:
+/*
+- documentation!!
+- where should the parameter event subscription go, and where should the publisher go
+
+- TODO for the long future: caching in client and service
+*/
+
+/// Internal rcl implementation struct
+struct rcl_parameter_client_impl_t;
+
+/// There is no sync/async parameter client distinction in rcl.
+typedef struct rcl_parameter_client_t
+{
+  // the parameter client provides storage and utility functions for parameter event subscription
+  // Should instead have a create_parameter_event_subscription function for this parameter client?
+  // rcl_subscription_t * parameter_event_subscription;
```

Comment:
Naively, I'd say that the fact that a subscription is involved is an implementation detail of parameters. A related question is whether or not the user can get a handle on the subscription object that the client uses or change the QoS options of that subscription.

If we don't expose these "details" of how parameters work, then I think it's clear that the subscription should be created and destroyed within `rcl` without extra action from the client library.

I think you've already done the other necessary bits to "abstract" the implementation of the parameter client and server from the user of `rcl` with `rcl_wait_set_add_parameter_client` and the matching `take` functions.

More to your question, can you describe in more detail what `create_parameter_event_subscription` would do and what the alternative default behavior is?